test/core/security/alts_credentials_fuzzer.cc

```diff
@@ -0,0 +1,149 @@
+      str = static_cast<char*>(gpr_realloc(str, cap));
+    }
+    c = static_cast<char>(read_byte(inp));
+    str[sz++] = c;
+  } while (c != 0);
+  return str;
+}
+
+// Add a random number of target service accounts to client options.
+static void read_target_service_accounts(
+    input_stream* inp, grpc_alts_credentials_options* options) {
+  size_t n = read_byte(inp);
+  for (size_t i = 0; i < n; i++) {
+    char* service_account = read_string(inp);
+    if (service_account != nullptr) {
+      grpc_alts_credentials_client_options_add_target_service_account(
+          options, service_account);
+      gpr_free(service_account);
+    }
+  }
+  // Added to improve code coverage.
+  grpc_alts_credentials_client_options_add_target_service_account(options,
+                                                                  nullptr);
+  grpc_alts_credentials_client_options_add_target_service_account(
+      nullptr, "this is service account");
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  char* grpc_trace_fuzzer = gpr_getenv("GRPC_TRACE_FUZZER");
+  if (squelch && grpc_trace_fuzzer == nullptr) {
+    gpr_set_log_function(dont_log);
+  }
+  gpr_free(grpc_trace_fuzzer);
+  struct grpc_memory_counters counters;
+  if (leak_check) {
+    grpc_memory_counters_init();
+  }
+  input_stream inp = {data, data + size};
+  grpc_init();
+  bool is_on_gcp = grpc_alts_is_running_on_gcp();
+  while (inp.cur != inp.end) {
+    bool enable_untrusted_alts = read_byte(&inp) & 0x01;
+    char* handshaker_service_url =
+        read_byte(&inp) & 0x01 ? read_string(&inp) : nullptr;
+    switch (read_byte(&inp)) {
+      // Test ALTS channel credentials.
+      case 0: {
+        grpc_alts_credentials_options* options =
+            grpc_alts_credentials_client_options_create();
+        read_target_service_accounts(&inp, options);
+        grpc_channel_credentials* cred =
+            grpc_alts_credentials_create_customized(
+                options, handshaker_service_url, enable_untrusted_alts);
+        if (!enable_untrusted_alts && !is_on_gcp) {
+          GPR_ASSERT(cred == nullptr);
+        } else {
+          GPR_ASSERT(cred != nullptr);
+        }
+        grpc_channel_credentials_release(cred);
+        grpc_alts_credentials_options_destroy(options);
+        break;
+      }
+      // Test ALTS server credentials.
+      case 1: {
+        grpc_alts_credentials_options* options =
+            grpc_alts_credentials_server_options_create();
+        grpc_server_credentials* cred =
+            grpc_alts_server_credentials_create_customized(
+                options, handshaker_service_url, enable_untrusted_alts);
+        if (!enable_untrusted_alts && !is_on_gcp) {
+          GPR_ASSERT(cred == nullptr);
+        } else {
+          GPR_ASSERT(cred != nullptr);
+        }
+        grpc_server_credentials_release(cred);
+        grpc_alts_credentials_options_destroy(options);
+        break;
+      }
+      default:
```

Comment:
It seems that most of cases where read_byte() does not produce 0 or 1, it will get into default case. It is not efficient. Why not read_bit, so that it guarantees it will go into either test channel credentials or test server credentials?

Reply:
Good points. The code is revised. PTAL